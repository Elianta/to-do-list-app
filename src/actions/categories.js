import {
    ADD_CATEGORY,
    ADD_NESTED_CATEGORY,
    EDIT_CATEGORY,
    REMOVE_CATEGORY,
    ADD_TASK,
    TOGGLE_TASK,
    SAVE_EDITED_TASK_WITH_TRANSFER,
    SAVE_EDITED_TASK_WITHOUT_TRANSFER
} from '../variables/actions';

let categoryCounter = 1;
let taskCounter = 1;

// ADD_CATEGORY
export const addCategory = (name) => ({
    type: ADD_CATEGORY,
    category: {
        id: categoryCounter++,
        name,
        children: [],
        tasks: []
    }
});

// ADD_NESTED_CATEGORY
export const addNestedCategory = (name, parentID) => ({
    type: ADD_NESTED_CATEGORY,
    parentID,
    category: {
        id: categoryCounter++,
        name,
        children: [],
        tasks: []
    }
});

// EDIT_CATEGORY
export const editCategory = (id, update) => ({
    type: EDIT_CATEGORY,
    id,
    update
});

// REMOVE_CATEGORY
export const removeCategory = (id) => ({
    type: REMOVE_CATEGORY,
    id
});

// ADD_TASK
export const addTask = (name, categoryID) => ({
    type: ADD_TASK,
    task: {
        id: taskCounter++,
        name,
        description: '',
        isDone: false
    },
    categoryID
});

// TOGGLE_TASK
export const toggleTask = (taskID, categoryID) => ({
    type: TOGGLE_TASK,
    taskID,
    categoryID
});

// SAVE_EDITED_TASK_WITHOUT_TRANSFER
export const saveEditedTaskWithoutTransfer = (task, categoryID) => ({
    type: SAVE_EDITED_TASK_WITHOUT_TRANSFER,
    task,
    categoryID
});

// SAVE_EDITED_TASK_WITH_TRANSFER
export const saveEditedTaskWithTransfer = (task, categoryID, destination) => ({
    type: SAVE_EDITED_TASK_WITH_TRANSFER,
    task,
    categoryID,
    destination
});
