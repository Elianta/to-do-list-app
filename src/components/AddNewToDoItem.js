import React from "react";

export default class AddNewToDoItem extends React.Component {
  constructor(props) {
    super(props);
    this.handleAddNewToDoItem = this.handleAddNewToDoItem.bind(this);

    this.state = {
      error: undefined
    };
  }

  handleAddNewToDoItem(e) {
    e.preventDefault();
    const toDoItem = e.target.elements.toDoItem.value.trim();
    const error = this.props.handleAddNewToDoItem(this.props.categoryID, toDoItem);

    this.setState(() => ({ error }));

    if (!error) {
      e.target.elements.toDoItem.value = '';
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form className="add-item-form" onSubmit={this.handleAddNewToDoItem}>
          <input className="add-item-form__input" type="text" name="toDoItem" placeholder="Enter item title"/>
          <button className="button">Add</button>
        </form>
      </div>
    )
  }
}
