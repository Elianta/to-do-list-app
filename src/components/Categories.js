import React from 'react';
import Category from './Category.js';

class Categories extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            categoryToMoveTaskInId: null
        };
        this.handleSpecifyCategoryToMove = this.handleSpecifyCategoryToMove.bind(this);
    }

    handleSpecifyCategoryToMove(categoryID) {
        this.setState(() => ({
            categoryToMoveTaskInId: categoryID
        }));
        this.props.handleSpecifyCategoryToMove(categoryID);
    }

    render() {
        let categories = this.props.categories.map((category) => {
            return (
                <Category
                    node={category}
                    children={category.children}
                    handleAddNestedCategory={this.props.handleAddNestedCategory}
                    key={category.name}
                    handleEditCategory={this.props.handleEditCategory}
                    handleDeleteCategory={this.props.handleDeleteCategory}
                    mode={this.props.mode}
                    updateProgressBar={this.props.updateProgressBar}
                    categoryID={this.props.categoryID}
                    handleSpecifyCategoryToMove={this.handleSpecifyCategoryToMove}
                    categoryToMoveTaskInId={this.state.categoryToMoveTaskInId}
                    transferTaskIn={this.state.categoryToMoveTaskInId === category.id}
                />
            );
        });

        return (
            <div className="categories__container">
                <ul className="categories__list">
                    {categories}
                </ul>
            </div>
        );
    }
}

export default Categories;
