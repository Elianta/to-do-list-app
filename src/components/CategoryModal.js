import React from 'react';
import Modal from 'react-modal';

const CategoryModal = (props) => (
    <Modal
        isOpen={props.categoryModalDisplayed}
        onRequestClose={props.handleModalClose}
        contentLabel="Selected Category"
        closeTimeoutMS={200}
        className="modal"
    >
        <h3 className="modal__title">{props.title}</h3>
        <form className="modal__form" onSubmit={props.handleSaveCategory}>
            <input
                type="text"
                name="category"
                autoFocus={true}
                className="modal__input"
                defaultValue={props.modalMode === 'edit' ? props.categoryName : ''}
            />
            <button className="button  modal__btn">Save</button>
        </form>
    </Modal>
);

export default CategoryModal;
