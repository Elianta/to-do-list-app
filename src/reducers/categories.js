import {
    ADD_CATEGORY,
    ADD_NESTED_CATEGORY,
    EDIT_CATEGORY,
    REMOVE_CATEGORY,
    ADD_TASK,
    TOGGLE_TASK,
    SAVE_EDITED_TASK_WITH_TRANSFER,
    SAVE_EDITED_TASK_WITHOUT_TRANSFER
} from '../variables/actions';

const categoriesReducer = (state = [], action) => {
    switch (action.type) {
        case ADD_CATEGORY:
            return [
                action.category,
                ...state
            ];
        case ADD_NESTED_CATEGORY:
            return addNestedCategory(state, action.parentID, action.category);
        case EDIT_CATEGORY:
            return editCategory(state, action.id, action.update);
        case REMOVE_CATEGORY:
            return deleteCategory(state, action.id);
        case ADD_TASK:
            return addTaskToCategories(state, action.categoryID, action.task);
        case TOGGLE_TASK:
            return toggleTask(state, action.taskID, action.categoryID);
        case SAVE_EDITED_TASK_WITHOUT_TRANSFER:
            return saveTaskWithoutTransfer(state, action.categoryID, action.task);
        case SAVE_EDITED_TASK_WITH_TRANSFER:
            return saveTaskWithTransfer(state, action.categoryID, action.task, action.destination);
        default:
            return state;
    }
};

export default categoriesReducer;

function addNestedCategory(categories, parentID, childCategory) {
    return categories.map((category) => {
        if (category.id === parentID) {
            category.children.unshift(childCategory);
        } else if (category.children.length) {
            addNestedCategory(category.children, parentID, childCategory);
        }
        return category;
    });
}

function deleteCategory(categories, id) {
    return categories.reduce((finalListOfCategories, category) => {
        if (category.id !== id) {
            if (category.children.length) {
                category.children = deleteCategory(category.children, id);
            }
            finalListOfCategories.push(category);
        }
        return finalListOfCategories;
    }, []);
}

function editCategory(categories, id, text) {
    return categories.map(function (category) {
        if (category.id === id) {
            category.name = text;
        } else if (category.children.length) {
            editCategory(category.children, id, text);
        }
        return category;
    });
}

function addTaskToCategories(categories, categoryID, task) {
    return categories.map((category) => {
        if (category.id === categoryID) {
            category.tasks.unshift(task);
        } else if (category.children.length) {
            addTaskToCategories(category.children, categoryID, task);
        }
        return category;
    });
}

function toggleTask(categories, taskID, categoryID) {
    return categories.map((category) => {
        if (category.id === categoryID) {
            category.tasks.forEach((item) => {
                if (item.id === taskID) {
                    item.isDone = !item.isDone;
                }
            });
        } else if (category.children.length) {
            toggleTask(category.children, taskID, categoryID);
        }
        return category;
    });
}

function saveTaskWithoutTransfer(categories, categoryID, task) {
    return categories.map((category) => {
        if (category.id === categoryID) {
            category.tasks.forEach((item) => {
                if (item.id === task.id) {
                    item.name = task.name;
                    item.description = task.description;
                    item.isDone = task.isDone;
                }
            });
        } else if (category.children.length) {
            saveTaskWithoutTransfer(category.children, categoryID, task);
        }
        return category;
    });
}

function saveTaskWithTransfer(categories, categoryID, task, destinationID) {
    return categories.map((category) => {
        if (category.id === categoryID) {
            category.tasks = category.tasks.filter((item) => item.id !== task.id);
        } else if (category.id === destinationID) {
            category.tasks.unshift(task);
        }
        if (category.children.length) {
            saveTaskWithTransfer(category.children, categoryID, task, destinationID);
        }
        return category;
    });
}
