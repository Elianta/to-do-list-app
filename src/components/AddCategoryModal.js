import React from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import Modal from 'react-modal';
import {closeModals} from '../actions/modals';
import {addNestedCategory} from '../actions/categories';

export class AddCategoryModal extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            category: ''
        };
        this.handleSaveNewCategory = this.handleSaveNewCategory.bind(this);
    }

    handleSaveNewCategory(e) {
        e.preventDefault();
        const categoryName = this.state.category;
        this.props.addNestedCategory({name: categoryName, parentID: this.props.modals.editableCategory.id});
        this.props.closeModals();
    }

    render() {
        return (
            <Modal
                isOpen={this.props.modals.isAddCategoryModalOpen}
                onRequestClose={this.props.handleModalClose}
                contentLabel="Add Category"
                closeTimeoutMS={200}
                className="modal"
            >
                <h3 className="modal__title">{`Add nested category into '${this.props.modals.editableCategory.name}'`}</h3>
                <form data-testid="form" className="modal__form" onSubmit={this.handleSaveNewCategory}>
                    <input
                        type="text"
                        name="category"
                        autoFocus={true}
                        className="modal__input"
                        data-testid="name"
                        onChange={(e) => {
                            this.setState({category: e.currentTarget.value.trim()});
                        }}
                    />
                    <button className="button  modal__btn">Save</button>
                </form>
            </Modal>
        );
    }
}

const mapStateToProps = (state) => ({
    modals: state.modals
});

const mapDispatchToProps = (dispatch) => {
    return bindActionCreators({
        addNestedCategory,
        closeModals
    }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(AddCategoryModal);
