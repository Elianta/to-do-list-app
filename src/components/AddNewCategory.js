import React from 'react';
import {connect} from 'react-redux';
import {addCategory} from '../actions/categories';

class AddNewCategory extends React.Component {
    constructor(props) {
        super(props);
        this.handleAddNewCategory = this.handleAddNewCategory.bind(this);
        this.state = {
            error: undefined
        };
    }

    handleAddNewCategory(e) {
        e.preventDefault();
        const categoryName = e.target.elements.category.value.trim();
        if (!categoryName) {
            this.setState(() => ({error: 'Enter valid value to add item'}));
        } else {
            this.setState(() => ({
                error: ''
            }));
            this.props.dispatch(addCategory(categoryName));
            e.target.elements.category.value = '';
        }
    }

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form className="add-item-form" onSubmit={this.handleAddNewCategory}>
                    <input
                        className="add-item-form__input"
                        type="text"
                        name="category"
                        placeholder="Enter category title..."
                    />
                    <button className="button">Add</button>
                </form>
            </div>
        );
    }
}

export default connect()(AddNewCategory);
