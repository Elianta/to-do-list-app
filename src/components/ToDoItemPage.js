import React from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import Categories from './Categories.js';
import {Link} from 'react-router-dom';
import {saveEditedTaskWithoutTransfer, saveEditedTaskWithTransfer} from '../actions/categories';

export class ToDoItemPage extends React.Component {
    constructor(props) {
        super(props);
        this.categoryID = parseInt(props.match.params.categoryID, 10);
        this.taskID = parseInt(props.match.params.taskID, 10);
        this.task = this.__findTaskByIds(props.categories, this.categoryID, this.taskID)[0];
        this.state = {
            isDone: this.task.isDone,
            name: this.task.name,
            description: this.task.description,
            categoryToMove: this.categoryID
        };
        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleSpecifyCategoryToMove = this.handleSpecifyCategoryToMove.bind(this);
    }

    handleChange(event) {
        const target = event.target;
        const value = target.type === 'checkbox' ? target.checked : target.value;
        const name = target.name;
        this.setState({
            [name]: value
        });
    }

    handleSubmit() {
        let {isDone, name, description} = this.state;
        const id = this.taskID;
        const taskData = {isDone, name, description, id};
        if (this.categoryID === this.state.categoryToMove) {
            this.props.saveEditedTaskWithoutTransfer(taskData, this.categoryID);
        } else {
            this.props.saveEditedTaskWithTransfer(taskData, this.categoryID, this.state.categoryToMove);
        }
    }

    handleSpecifyCategoryToMove(categoryID) {
        this.setState(() => ({
            categoryToMove: categoryID
        }));
    }

    __findTaskByIds(categories, categoryID, taskID) {
        return categories.reduce((prev, category) => {
            if (category.id === categoryID) {
                let result = category.tasks.filter((task) => {
                    return task.id === taskID;
                });
                return prev.concat(result);
            } else if (category.children.length && !prev.length) {
                return this.__findTaskByIds(category.children, categoryID, taskID);
            }
            return prev;
        }, []);
    }

    render() {
        return (
            <div className="page-content">
                <header className="header">
                    <div className="container  container--horizontal">
                        <h1 className="header__title">{this.state.name}</h1>
                    </div>
                </header>
                <main>
                    <div className="container  container--horizontal">
                        <section className="categories">
                            <Categories
                                active={this.props.match.params.categoryID}
                                mode="taskTransfer"
                                categoryID={this.categoryID}
                                handleSpecifyCategoryToMove={this.handleSpecifyCategoryToMove}
                            />
                        </section>
                        <form className="todoitem">
                            <div className="todoitem__manage">
                                <Link
                                    data-testid="submit-btn"
                                    className="button  todoitem__btn"
                                    to={`/category/${this.categoryID}`}
                                    onClick={this.handleSubmit}
                                >Save Changes</Link>
                                <Link
                                    data-testid="cancel-btn"
                                    className="button  todoitem__btn"
                                    to={`/category/${this.categoryID}`}
                                >Cancel</Link>
                            </div>
                            <input
                                data-testid="name"
                                className="todoitem__title"
                                type="text"
                                name="name"
                                value={this.state.name}
                                autoFocus={true}
                                onChange={this.handleChange}
                            />
                            <div className="option">
                                <input
                                    data-testid="is-done"
                                    className="option__input"
                                    type="checkbox"
                                    id="taskDone"
                                    name="isDone"
                                    checked={this.state.isDone}
                                    onChange={this.handleChange}
                                />
                                <label className="option__label" htmlFor="taskDone">Done</label>
                            </div>

                            <label className="todoitem__description-label" htmlFor="description">Description:</label>
                            <textarea
                                data-testid="description"
                                className="todoitem__description"
                                placeholder="Write something..."
                                id="description"
                                name="description"
                                value={this.state.description}
                                onChange={this.handleChange}
                            />

                        </form>
                    </div>
                </main>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    categories: state.categories
});

const mapDispatchToProps = (dispatch) => {
    return bindActionCreators({
        saveEditedTaskWithoutTransfer,
        saveEditedTaskWithTransfer
    }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(ToDoItemPage);
