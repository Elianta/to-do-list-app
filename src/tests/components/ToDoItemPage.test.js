import React from 'react';
import {shallow} from 'enzyme';
import {ToDoItemPage} from '../../components/ToDoItemPage';
import categories from '../fixtures/categories';

const match = {
    params: {
        categoryID: '5',
        taskID: '2'
    }
};


test('should render ToDoItemPage correctly', () => {
    const wrapper = shallow(<ToDoItemPage match={match} categories={categories}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should search task by id correctly and write data in state', () => {
    const task2category5 = categories[0].children[1].children[0].tasks[1];
    const wrapper = shallow(<ToDoItemPage match={match} categories={categories}/>);
    expect(wrapper.instance().task).toEqual(task2category5);
});

describe('test fill in the form', () => {
    let wrapper;
    const newName = 'Changed Task Name';
    const newDescription = 'Changed Task Description';
    const newIsDone = false;

    beforeEach(() => {
        const saveEditedTaskWithoutTransfer = jest.fn();
        const saveEditedTaskWithTransfer = jest.fn();
        wrapper = shallow(
            <ToDoItemPage
                match={match}
                categories={categories}
                saveEditedTaskWithoutTransfer={saveEditedTaskWithoutTransfer}
                saveEditedTaskWithTransfer={saveEditedTaskWithTransfer}/>
        );
        const name = wrapper.find('[data-testid="name"]');
        const isDone = wrapper.find('[data-testid="is-done"]');
        const description = wrapper.find('[data-testid="description"]');
        name.simulate('change', {target: {value: newName, type: 'text', name: 'name'}});
        isDone.simulate('change', {target: {checked: newIsDone, type: 'checkbox', name: 'isDone'}});
        description.simulate('change', {target: {value: newDescription, type: 'text', name: 'description'}});
    });

    test('should send correct data when save task WITHOUT transfer to another category', () => {
        const submitBtn = wrapper.find('[data-testid="submit-btn"]');
        submitBtn.simulate('click');
        expect(wrapper.instance().props.saveEditedTaskWithoutTransfer).toHaveBeenCalledWith(
            {
                isDone: newIsDone,
                name: newName,
                description: newDescription,
                id: +match.params.taskID
            },
            +match.params.categoryID
        );
        expect(wrapper.instance().props.saveEditedTaskWithTransfer).not.toHaveBeenCalled();
    });

    test('should send correct data when save task WITH transfer to another category', () => {
        const submitBtn = wrapper.find('[data-testid="submit-btn"]');
        const categoryToMove = 1;
        wrapper.setState({categoryToMove: categoryToMove});
        submitBtn.simulate('click');
        expect(wrapper.instance().props.saveEditedTaskWithTransfer).toHaveBeenCalledWith(
            {
                isDone: newIsDone,
                name: newName,
                description: newDescription,
                id: +match.params.taskID
            },
            +match.params.categoryID,
            categoryToMove
        );
        expect(wrapper.instance().props.saveEditedTaskWithoutTransfer).not.toHaveBeenCalled();
    });

    test('should send NO data on cancel', () => {
        const cancelBtn = wrapper.find('[data-testid="cancel-btn"]');
        cancelBtn.simulate('click');
        expect(wrapper.instance().props.saveEditedTaskWithTransfer).not.toHaveBeenCalled();
        expect(wrapper.instance().props.saveEditedTaskWithoutTransfer).not.toHaveBeenCalled();
    });
});
