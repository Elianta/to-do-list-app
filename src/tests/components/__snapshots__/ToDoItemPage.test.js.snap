// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ToDoItemPage correctly 1`] = `
<div
  className="page-content"
>
  <header
    className="header"
  >
    <div
      className="container  container--horizontal"
    >
      <h1
        className="header__title"
      >
        Task 5
      </h1>
    </div>
  </header>
  <main>
    <div
      className="container  container--horizontal"
    >
      <section
        className="categories"
      >
        <Connect(Categories)
          active="5"
          categoryID={5}
          handleSpecifyCategoryToMove={[Function]}
          mode="taskTransfer"
        />
      </section>
      <form
        className="todoitem"
      >
        <div
          className="todoitem__manage"
        >
          <Link
            className="button  todoitem__btn"
            data-testid="submit-btn"
            onClick={[Function]}
            replace={false}
            to="/category/5"
          >
            Save Changes
          </Link>
          <Link
            className="button  todoitem__btn"
            data-testid="cancel-btn"
            replace={false}
            to="/category/5"
          >
            Cancel
          </Link>
        </div>
        <input
          autoFocus={true}
          className="todoitem__title"
          data-testid="name"
          name="name"
          onChange={[Function]}
          type="text"
          value="Task 5"
        />
        <div
          className="option"
        >
          <input
            checked={true}
            className="option__input"
            data-testid="is-done"
            id="taskDone"
            name="isDone"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="option__label"
            htmlFor="taskDone"
          >
            Done
          </label>
        </div>
        <label
          className="todoitem__description-label"
          htmlFor="description"
        >
          Description:
        </label>
        <textarea
          className="todoitem__description"
          data-testid="description"
          id="description"
          name="description"
          onChange={[Function]}
          placeholder="Write something..."
          value="Task 5 Text"
        />
      </form>
    </div>
  </main>
</div>
`;
